lib/registry-secrets.ts

```diff
@@ -0,0 +1,95 @@
+/**
+ * @license
+ * Copyright 2018 Balena Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * This module contains interfaces and classes for the validation of the JSON
+ * format expected by the balena builder and the docker daemon for the
+ * authentication of private docker registries.
+ */
+
+import * as ajv from 'ajv';
+
+import { RegistrySecretValidationError } from './errors';
+export { RegistrySecretValidationError } from './errors';
+
+export interface RegistrySecrets {
+	[registryAddress: string]: {
+		username: string;
+		password: string;
+	};
+}
+
+/**
+ * JSON schema validator for the private registry "secrets" (username and
+ * password entries keyed by hostname:port registry addresses).
+ */
+export class RegistrySecretValidator {
+	private registrySecretJsonSchema = {
+		// Sample valid registrySecrets JSON contents:
+		//   {  "docker.example.com": {"username": "ann", "password": "hunter2"},
+		//      "https://idx.docker.io/v1/": {"username": "mck", "password": "cze14"}
+		//   }
+		type: 'object',
+		patternProperties: {
+			'^\\S+$': {
+				type: 'object',
+				properties: {
+					username: { type: 'string' },
+					password: { type: 'string' },
+				},
+				additionalProperties: false,
+			},
+		},
+		additionalProperties: false,
+	};
+	private validator: ajv.Ajv;
+	private validateFunction: ajv.ValidateFunction;
+
+	constructor() {
+		this.validator = new ajv();
```

Comment:
I'd prefer having these at module level and avoiding constant recompilation, but if you are doing it as part of the class initialisation then you can just write it as:
```ts
private validator = new ajv()
private validateFunction = this.validator.compile(...)
```
and the types can be automatically picked up as well

Reply:
I've moved the initialisation to class scope instead of the constructor as suggested, thanks, but as for:

> _I'd prefer having these at module level and avoiding constant recompilation_

I think this ends up being a choice of functional versus OO programming style, with implications on what gets executed as a "side effect" of merely importing the module.

I agree that schema recompilation is wasteful and should be avoided, but the OO approach to it is to cache an instance of the validator object -- which is what I am doing in `resin-builder` `metadata.ts`. If you only create one object, you only do one compilation. In the actual use cases I made of this class in `balena-cli` and `resin-builder`, schema compilation only happens once in the process lifetime. Looking at the class' contents, if I move the `compile()` call to module level, then there is no reason for the class to exist and the methods should be moved to module level as well, and we have functional programming. Nothing "wrong" with functional programming, except for the experience I describe below...

Generally, running code at module level reduces flexibility and adds work to "mock it out" when writing unit tests. Suppose this module grows to contain a few classes in it, and the schema compilation is only relevant to one of the classes (`RegistrySecretValidator`). If we wanted to import the module to use one of the classes other than `RegistrySecretValidator` (whether for unit testing or production use), we might want to avoid any schema compilation at all (we'd want to run it zero times rather than running it "only once").

This schema compilation is fairly harmless to be done at module import time -- just some CPU usage. But in my previous job, the codebase was "90%" functional programming (in Python) with all sorts of initialisation taking place at module level / module import time: creating files, accessing databases, firing network heartbeats, downloading/uploading stuff to S3... When I joined, there were very few unit tests and many unresolved bugs. Management then made a decision that we should write unit tests to achieve a minimum of 70% code coverage. But writing unit tests for the functions in large modules with lots of module-level initialisation was a lot of effort to mock things out, to avoid the side effects on the filesystem, database and remote systems. So I can say that I got bruised by that experience :-) and now I am very conscious to avoid running anything at module level, if I can.

Now, if you look at what I did in `resin-builder` `metadata.ts`, I added these two lines:

```
// Cached at module level to speed up repeat validation
const registrySecretValidator = new RegistrySecretValidator();
```
I created the object at module level which is the opposite of what I am preaching here! :-) But I did that because the builder was implemented as a module-level `build()` function in `resin-builder` `run-build.ts`, in functional programming style without any `init_build()` sort of function to create and cache state. If it had been designed in OO style, a `Builder` class with a `build()` method, then I would have added `this.registrySecretValidator = new RegistrySecretValidator();` to the `Builder` class' constructor, and hopefully a long-lived build server process would create and cache an instance of `Builder` to avoid reinitialising several things, including schema compilation. I didn't look at it in depth, but the `build()` and `runBuild()` functions in `resin-builder` `run-build.ts` are very large and seem to initialise lots of things, some of which I suspect could have been cached in a long-lived builder process.